Instructions

In this checkpoint, we are going to work on the billing history of the Tunisian electricity and gas company and apply what you learned in pre-processing chapter.

Dataset description : 

The provided dataset contains the billing history of the Tunisian electricity and gas company's customers from 2005 to 2019.

➡️ Dataset link
https://i.imgur.com/AaBqRR5.png

1.	Load the dataset, display the ten first lines, store the results in a variable called 'client_0_bills'.

2.	What is the data type of the 'client_0_bills' variable ?

3.	Display the general information of the dataset and try to answer the following questions : 

o	How many rows and columns do we have in this dataset ?

o	How many categorical features are present in the dataset ?

o	How much memory space does the dataset consume ?

4.	Inspect the dataset for potential missing values.

5.	Select your strategy to handle missing values, and tell us why you had made that choice.

6.	Run a descriptive analysis on numeric features (columns).

7.	Select the bills records for the client with an id ='train_Client_0', using 2 methods.

8.	Transform the 'counter_type' feature to a numeric variable using the encoder of your choice.

9.	Delete the 'counter_statue' feature from the Dataframe

In [2]:
import pandas as pd

import numpy as np

In [4]:
df=pd.read_csv('STEG_BILLING_HISTORY.csv',low_memory=False)

df

client_id invoice_date  tarif_type  counter_number  \
0            train_Client_0   2014-03-24          11       1335667.0   
1            train_Client_0   2013-03-29          11       1335667.0   
2            train_Client_0   2015-03-23          11       1335667.0   
3            train_Client_0   2015-07-13          11       1335667.0   
4            train_Client_0   2016-11-17          11       1335667.0   
...                     ...          ...         ...             ...   
4476744  train_Client_99998   2005-08-19          10       1253571.0   
4476745  train_Client_99998   2005-12-19          10       1253571.0   
4476746  train_Client_99999   1996-09-25          11        560948.0   
4476747  train_Client_99999   1996-05-28          11        560948.0   
4476748  train_Client_99999   1996-01-25          11        560948.0   

        counter_statue  counter_code  reading_remarque  counter_coefficient  \
0                    0           203               8.0                    1   
1                    0           203               6.0                    1   
2                    0           203               8.0                    1   
3                    0           207               8.0                    1   
4                    0           207               9.0                    1   
...                ...           ...               ...                  ...   
4476744              0           202               9.0                    1   
4476745              0           202               6.0                    1   
4476746              0           203               6.0                    1   
4476747              0           203               6.0                    1   
4476748              0           203               6.0                    1   

         consommation_level_1  consommation_level_2  consommation_level_3  \
0                          82                     0                     0   
1                        1200                   184                     0   
2                         123                     0                     0   
3                         102                     0                     0   
4                         572                     0                     0   
...                       ...                   ...                   ...   
4476744                   400                   135                     0   
4476745                   200                     6                     0   
4476746                   259                     0                     0   
4476747                   603                     0                     0   
4476748                   516                     0                     0   

         consommation_level_4  old_index  new_index  months_number  \
0                           0      14302      14384              4   
1                           0      12294      13678              4   
2                           0      14624      14747              4   
3                           0      14747      14849              4   
4                           0      15066      15638             12   
...                       ...        ...        ...            ...   
4476744                     0       3197       3732              8   
4476745                     0       3732       3938              4   
4476746                     0      13884      14143              4   
4476747                     0      13281      13884              4   
4476748                     0      12765      13281              4   

        counter_type  
0               ELEC  
1               ELEC  
2               ELEC  
3               ELEC  
4               ELEC  
...              ...  
4476744         ELEC  
4476745         ELEC  
4476746         ELEC  
4476747         ELEC  
4476748         ELEC  

[4476749 rows x 16 columns]

In [28]:
# Q1

client_0_bills=df.head(10)

client_0_bills

client_id invoice_date  tarif_type  counter_number counter_statue  \
0  train_Client_0   2014-03-24          11       1335667.0              0   
1  train_Client_0   2013-03-29          11       1335667.0              0   
2  train_Client_0   2015-03-23          11       1335667.0              0   
3  train_Client_0   2015-07-13          11       1335667.0              0   
4  train_Client_0   2016-11-17          11       1335667.0              0   
5  train_Client_0   2017-07-17          11       1335667.0              0   
6  train_Client_0   2018-12-07          11       1335667.0              0   
7  train_Client_0   2019-03-19          11       1335667.0              0   
8  train_Client_0   2011-07-22          11       1335667.0              0   
9  train_Client_0   2011-11-22          11       1335667.0              0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203               8.0                    1                    82   
1           203               6.0                    1                  1200   
2           203               8.0                    1                   123   
3           207               8.0                    1                   102   
4           207               9.0                    1                   572   
5           207               9.0                    1                   314   
6           207               9.0                    1                   541   
7           207               9.0                    1                   585   
8           203               9.0                    1                  1200   
9           203               6.0                    1                  1082   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                     0                     0                     0   
1                   184                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   
5                     0                     0                     0   
6                     0                     0                     0   
7                     0                     0                     0   
8                   186                     0                     0   
9                     0                     0                     0   

   old_index  new_index  months_number counter_type  
0      14302      14384              4         ELEC  
1      12294      13678              4         ELEC  
2      14624      14747              4         ELEC  
3      14747      14849              4         ELEC  
4      15066      15638             12         ELEC  
5      15638      15952              8         ELEC  
6      15952      16493             12         ELEC  
7      16493      17078              8         ELEC  
8       7770       9156              4         ELEC  
9       9156      10238              4         ELEC

In [ ]:
# Q2 , its a dataframe

type(client_0_bills)

pandas.core.frame.DataFrame

In [29]:
# Q3 

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4476749 entries, 0 to 4476748
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   int64  
 8   consommation_level_1  int64  
 9   consommation_level_2  int64  
 10  consommation_level_3  int64  
 11  consommation_level_4  int64  
 12  old_index             int64  
 13  new_index             int64  
 14  months_number         int64  
 15  counter_type          object 
dtypes: float64(2), int64(10), object(4)
memory usage: 546.5+ MB


* we have 4 476 749 rows and 16 columns in our dataset

* we have 4 categorical features in our dataset

* the dataset is consuming 546.5+ KB memory space

In [10]:
# Q4

df.isnull().sum()

client_id                  0
invoice_date               0
tarif_type                 0
counter_number            48
counter_statue             0
counter_code               0
reading_remarque        4531
counter_coefficient        0
consommation_level_1       0
consommation_level_2       0
consommation_level_3       0
consommation_level_4       0
old_index                  0
new_index                  0
months_number              0
counter_type               0
dtype: int64

# **Q5**

# as *counter_number* and *reading_remarque* are both numerical data we can use either median or mode to replace missing values

# mean is excluded as it returns float result which will not help us in our case

In [11]:
df['counter_number'].fillna(df['counter_number'].median(),inplace=True)

C:\Users\t400862\AppData\Local\Temp\ipykernel_23368\3141265419.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['counter_number'].fillna(df['counter_number'].median(),inplace=True)


In [12]:
df['reading_remarque']=df['reading_remarque'].fillna(df['reading_remarque'].median())

In [13]:
df.isnull().sum()

client_id               0
invoice_date            0
tarif_type              0
counter_number          0
counter_statue          0
counter_code            0
reading_remarque        0
counter_coefficient     0
consommation_level_1    0
consommation_level_2    0
consommation_level_3    0
consommation_level_4    0
old_index               0
new_index               0
months_number           0
counter_type            0
dtype: int64

In [14]:
# Q6

df.describe()

tarif_type  counter_number  counter_code  reading_remarque  \
count  4.476749e+06    4.476749e+06  4.476749e+06      4.476749e+06   
mean   2.012804e+01    1.230582e+11  1.724884e+02      7.322400e+00   
std    1.347256e+01    1.657267e+12  1.338871e+02      1.571212e+00   
min    8.000000e+00    0.000000e+00  0.000000e+00      5.000000e+00   
25%    1.100000e+01    1.211080e+05  5.000000e+00      6.000000e+00   
50%    1.100000e+01    4.945610e+05  2.030000e+02      8.000000e+00   
75%    4.000000e+01    1.115161e+06  2.070000e+02      9.000000e+00   
max    4.500000e+01    2.798115e+13  6.000000e+02      4.130000e+02   

       counter_coefficient  consommation_level_1  consommation_level_2  \
count         4.476749e+06          4.476749e+06          4.476749e+06   
mean          1.003040e+00          4.109795e+02          1.093225e+02   
std           3.083466e-01          7.573080e+02          1.220123e+03   
min           0.000000e+00          0.000000e+00          0.000000e+00   
25%           1.000000e+00          7.900000e+01          0.000000e+00   
50%           1.000000e+00          2.740000e+02          0.000000e+00   
75%           1.000000e+00          6.000000e+02          0.000000e+00   
max           5.000000e+01          9.999100e+05          9.990730e+05   

       consommation_level_3  consommation_level_4     old_index     new_index  \
count          4.476749e+06          4.476749e+06  4.476749e+06  4.476749e+06   
mean           2.030620e+01          5.292588e+01  1.776700e+04  1.834970e+04   
std            1.574239e+02          8.754725e+02  4.036693e+04  4.095321e+04   
min            0.000000e+00          0.000000e+00  0.000000e+00  0.000000e+00   
25%            0.000000e+00          0.000000e+00  1.791000e+03  2.056000e+03   
50%            0.000000e+00          0.000000e+00  7.690000e+03  8.192000e+03   
75%            0.000000e+00          0.000000e+00  2.166000e+04  2.234300e+04   
max            6.449200e+04          5.479460e+05  2.800280e+06  2.870972e+06   

       months_number  
count   4.476749e+06  
mean    4.483095e+01  
std     3.128335e+03  
min     0.000000e+00  
25%     4.000000e+00  
50%     4.000000e+00  
75%     4.000000e+00  
max     6.366240e+05

In [15]:
# Q7

df.query("client_id=='train_Client_0'")

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0            203               8.0                    1                    82   
1            203               6.0                    1                  1200   
2            203               8.0                    1                   123   
3            207               8.0                    1                   102   
4            207               9.0                    1                   572   
5            207               9.0                    1                   314   
6            207               9.0                    1                   541   
7            207               9.0                    1                   585   
8            203               9.0                    1                  1200   
9            203               6.0                    1                  1082   
10           203               6.0                    1                   174   
11           203               6.0                    1                   628   
12           203               6.0                    1                   886   
13           203               8.0                    1           

In [30]:
df[df['client_id']=='train_Client_0']

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0            203               8.0                    1                    82   
1            203               6.0                    1                  1200   
2            203               8.0                    1                   123   
3            207               8.0                    1                   102   
4            207               9.0                    1                   572   
5            207               9.0                    1                   314   
6            207               9.0                    1                   541   
7            207               9.0                    1                   585   
8            203               9.0                    1                  1200   
9            203               6.0                    1                  1082   
10           203               6.0                    1                   174   
11           203               6.0                    1                   628   
12           203               6.0                    1                   886   
13           203               8.0                    1           

In [17]:
# Q8

df['counter_type'].unique()

array(['ELEC', 'GAZ'], dtype=object)

In [18]:
from sklearn .preprocessing import LabelEncoder

In [19]:
label_encoder=LabelEncoder()

df['counter_type']=label_encoder.fit_transform(df['counter_type'])

df['counter_type'].unique()

array([0, 1])

In [25]:
df['counter_type'].value_counts()

counter_type
0    3079406
1    1397343
Name: count, dtype: int64

In [31]:
# Q9

df

client_id invoice_date  tarif_type  counter_number  \
0            train_Client_0   2014-03-24          11       1335667.0   
1            train_Client_0   2013-03-29          11       1335667.0   
2            train_Client_0   2015-03-23          11       1335667.0   
3            train_Client_0   2015-07-13          11       1335667.0   
4            train_Client_0   2016-11-17          11       1335667.0   
...                     ...          ...         ...             ...   
4476744  train_Client_99998   2005-08-19          10       1253571.0   
4476745  train_Client_99998   2005-12-19          10       1253571.0   
4476746  train_Client_99999   1996-09-25          11        560948.0   
4476747  train_Client_99999   1996-05-28          11        560948.0   
4476748  train_Client_99999   1996-01-25          11        560948.0   

        counter_statue  counter_code  reading_remarque  counter_coefficient  \
0                    0           203               8.0                    1   
1                    0           203               6.0                    1   
2                    0           203               8.0                    1   
3                    0           207               8.0                    1   
4                    0           207               9.0                    1   
...                ...           ...               ...                  ...   
4476744              0           202               9.0                    1   
4476745              0           202               6.0                    1   
4476746              0           203               6.0                    1   
4476747              0           203               6.0                    1   
4476748              0           203               6.0                    1   

         consommation_level_1  consommation_level_2  consommation_level_3  \
0                          82                     0                     0   
1                        1200                   184                     0   
2                         123                     0                     0   
3                         102                     0                     0   
4                         572                     0                     0   
...                       ...                   ...                   ...   
4476744                   400                   135                     0   
4476745                   200                     6                     0   
4476746                   259                     0                     0   
4476747                   603                     0                     0   
4476748                   516                     0                     0   

         consommation_level_4  old_index  new_index  months_number  \
0                           0      14302      14384              4   
1                           0      12294      13678              4   
2                           0      14624      14747              4   
3                           0      14747      14849              4   
4                           0      15066      15638             12   
...                       ...        ...        ...            ...   
4476744                     0       3197       3732              8   
4476745                     0       3732       3938              4   
4476746                     0      13884      14143              4   
4476747                     0      13281      13884              4   
4476748                     0      12765      13281              4   

        counter_type  
0               ELEC  
1               ELEC  
2               ELEC  
3               ELEC  
4               ELEC  
...              ...  
4476744         ELEC  
4476745         ELEC  
4476746         ELEC  
4476747         ELEC  
4476748         ELEC  

[4476749 rows x 16 columns]

In [ ]:
df.drop(columns='counter_statue',inplace=True)

df

client_id invoice_date  tarif_type  counter_number  \
0            train_Client_0   2014-03-24          11       1335667.0   
1            train_Client_0   2013-03-29          11       1335667.0   
2            train_Client_0   2015-03-23          11       1335667.0   
3            train_Client_0   2015-07-13          11       1335667.0   
4            train_Client_0   2016-11-17          11       1335667.0   
...                     ...          ...         ...             ...   
4476744  train_Client_99998   2005-08-19          10       1253571.0   
4476745  train_Client_99998   2005-12-19          10       1253571.0   
4476746  train_Client_99999   1996-09-25          11        560948.0   
4476747  train_Client_99999   1996-05-28          11        560948.0   
4476748  train_Client_99999   1996-01-25          11        560948.0   

         counter_code  reading_remarque  counter_coefficient  \
0                 203               8.0                    1   
1                 203               6.0                    1   
2                 203               8.0                    1   
3                 207               8.0                    1   
4                 207               9.0                    1   
...               ...               ...                  ...   
4476744           202               9.0                    1   
4476745           202               6.0                    1   
4476746           203               6.0                    1   
4476747           203               6.0                    1   
4476748           203               6.0                    1   

         consommation_level_1  consommation_level_2  consommation_level_3  \
0                          82                     0                     0   
1                        1200                   184                     0   
2                         123                     0                     0   
3                         102                     0                     0   
4                         572                     0                     0   
...                       ...                   ...                   ...   
4476744                   400                   135                     0   
4476745                   200                     6                     0   
4476746                   259                     0                     0   
4476747                   603                     0                     0   
4476748                   516                     0                     0   

         consommation_level_4  old_index  new_index  months_number  \
0                           0      14302      14384              4   
1                           0      12294      13678              4   
2                           0      14624      14747              4   
3                           0      14747      14849              4   
4                           0      15066      15638             12   
...                       ...        ...        ...            ...   
4476744                     0       3197       3732              8   
4476745                     0       3732       3938              4   
4476746                     0      13884      14143              4   
4476747                     0      13281      13884              4   
4476748                     0      12765      13281              4   

         counter_type  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
4476744             0  
4476745             0  
4476746             0  
4476747             0  
4476748             0  

[4476749 rows x 15 columns]